File: app/src/main/java/com/example/pruebaconceptonavigationmanager/actions/ActionMapperImpl.java
package com.example.pruebaconceptonavigationmanager.actions;

import android.content.Context;

import com.example.navigation.action.Action;
import com.example.navigation.action.ActionMapper;
import com.example.navigation.action.Field;
import com.example.navigation.action.RuleAction;
import com.example.pruebaconceptonavigationmanager.calculator.CalculatorActivity;
import com.example.pruebaconceptonavigationmanager.chooser.ChooserActivity;
import com.example.pruebaconceptonavigationmanager.description.DescriptionActivity;
import com.example.pruebaconceptonavigationmanager.installments.InstallmentsActivity;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import java.util.List;

//@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "name")
//@JsonSubTypes({
//        @JsonSubTypes.Type(value = CalculatorActivity.class, name = "CALCULATOR"),
//        @JsonSubTypes.Type(value = DescriptionActivity.class, name = "CALCULATOR"),
//        @JsonSubTypes.Type(value = ChooserActivity.class, name = "CHOOSER"),
//        @JsonSubTypes.Type(value = InstallmentsActivity.class, name = "INSTALLMENTS")})

public class ActionMapperImpl implements ActionMapper {

    Context context;

    public ActionMapperImpl(Context context) {
        this.context = context;
    }


    /**
     * Inicia una instancia de Action a partir de un deep link
     * @param action nodo que representa un action del json
     * @param params
     */
    @Override
    public void executeAction(RuleAction action, Object... params) {

        String name = action.getName();

        String deepLink = getDeepLink(name);

        initAction(deepLink, params);

    }
    /**
     * Sobre una Action ya activa, ejecuta en base a nuevos fields requeridos por otro step
     * @param action  Action activa (pantalla)
     * @param params
     */
    @Override
    public void executeNextField(Action action, Object... params) {

        String actionName = action.getName();

        action.execute((List<Field>)params[0]);


    }

    private String getDeepLink(String actionName){
        //TODO: Buscar el deep link asociado
        return "";
    }

    private void initAction(String deepLink, Object...params){
       //TODO: intsanciar con deep link
    }


}

File: app/src/main/java/com/example/pruebaconceptonavigationmanager/actions/ActionMapperImpl.kt
package com.example.pruebaconceptonavigationmanager.actions

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.example.navigation.action.Action
import com.example.navigation.action.ActionMapper
import com.example.navigation.action.BUNDLE_FIELDS
import com.example.navigation.action.RuleAction
import com.example.navigation.stepsEngine.field.Field
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategy


class ActionMapperImpl(var context: Context) : ActionMapper {

    private val INTERNAL = "INTERNAL"
    var deepLinks : DeepLinkMapping

    init {
        val mapper = ObjectMapper()
        mapper.propertyNamingStrategy = PropertyNamingStrategy.SNAKE_CASE
        val inputStreamActions= context.assets.open("actionsMapping/MLA.json")
        deepLinks = mapper.readValue(inputStreamActions, object : TypeReference<DeepLinkMapping>() {})
    }
    /**
     * Inicia una instancia de Action a partir de un deep link
     * @param action nodo que representa un action del json
     * @param params
     */
    override fun startAction(action: RuleAction) {
        val name = action.id
        val deepLink = getDeepLink(name)
        initAction(deepLink!!, action.getFields())
    }

    /**
     * Sobre una Action ya activa, ejecuta en base a nuevos fields requeridos por otro step
     * @param action  Action activa (pantalla)
     * @param params
     */
    override fun executeNextField(action: Action, vararg params: Any?) {
        action.executeFields(params[0] as List<Field>)
    }

    private fun getDeepLink(actionName: String?): String? {
        return deepLinks.actions!![actionName]
    }

    private fun initAction(deepLink: String, fields: ArrayList<Field>?) {
        val uri = Uri.parse(deepLink)
        val intent = getSafeIntent(context)
        intent!!.data = uri
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        intent.putParcelableArrayListExtra(BUNDLE_FIELDS, fields)
        context.startActivity(intent)
    }

    private fun getSafeIntent(context: Context): Intent? {
        return Intent(Intent.ACTION_VIEW).setPackage(context.packageName).putExtra(INTERNAL, true)
    }

}